docker-entrypoint.sh

```diff
@@ -54,60 +54,58 @@ if [[ ! -f /db/init_done ]] ; then
-              cat /db/cookie.jar
-              exit
-              ;;
-            *)
-              echo "Failed to download planet file. HTTP status code: ${CURL_STATUS_CODE}"
-              cat /db/planet.osm.bz2
-              exit
-              ;;
-          esac
-          exit
+        CURL_STATUS_CODE=$(curl -L -b /db/cookie.jar -o /db/planet.osm.bz2 -w "%{http_code}" "${OVERPASS_PLANET_URL}")
+        # try again until it's allowed
+        while [ $CURL_STATUS_CODE = "429" ] ; do
+            echo "Server responded with 429 Too many requests. Trying again in 5 minutes..."
+            sleep 300
+            CURL_STATUS_CODE=$(curl -L -b /db/cookie.jar -o /db/planet.osm.bz2 -w "%{http_code}" "${OVERPASS_PLANET_URL}")
         done
+          # for `file:///` scheme curl returns `000` HTTP status code
+        if [[ $CURL_STATUS_CODE = "200" || $CURL_STATUS_CODE = "000" ]]; then
+            (
+              if [[ ! -z "${OVERPASS_PLANET_PREPROCESS+x}" ]]; then
+                  echo "Running preprocessing command: ${OVERPASS_PLANET_PREPROCESS}"
+                  eval "${OVERPASS_PLANET_PREPROCESS}"
+              fi \
+              && /app/bin/init_osm3s.sh /db/planet.osm.bz2 /db/db /app "${META}" "--version=$(osmium fileinfo -e -g data.timestamp.last /db/planet.osm.bz2) --compression-method=${OVERPASS_COMPRESSION} --map-compression-method=${OVERPASS_COMPRESSION} --flush-size=${OVERPASS_FLUSH_SIZE}" \
+              && echo "Database created. Now updating it." \
+              && cp -r /app/etc/rules /db/db \
+              && chown -R overpass:overpass /db \
+              && echo "Updating" \
+              && /app/bin/update_overpass.sh "-O /db/planet.osm.bz2" \
+              && if [[ "OVERPASS_USE_AREAS" = "true" ]]; then
+                   echo "Generating areas..." && /app/bin/osm3s_query --progress --rules --db-dir=/db/db < /db/db/rules/areas.osm3s
+                 fi \
+              && touch /db/init_done \
+              && rm /db/planet.osm.bz2 \
+              && chown -R overpass:overpass /db \
+              && echo "Overpass container ready to receive requests"
+            ) || (
+              echo "Failed to process planet file"
+              exit 1
+            )
+        elif [[ $CURL_STATUS_CODE = "403" ]]; then
+          echo "Access denied when downloading planet file. Check your OVERPASS_PLANET_URL and OVERPASS_COOKIE_JAR_CONTENTS or USE_OAUTH_COOKIE_CLIENT"
+          cat /db/cookie.jar
+          exit 1
+        else
+          echo "Failed to download planet file. HTTP status code: ${CURL_STATUS_CODE}"
+          cat /db/planet.osm.bz2
+          exit 1
+        fi
     fi
 fi
 
 envsubst '${OVERPASS_MAX_TIMEOUT}' < /etc/nginx/nginx.conf.template > /etc/nginx/nginx.conf
 
 echo "Starting supervisord process"
 exec /usr/bin/supervisord -c /etc/supervisor/conf.d/supervisord.conf
+
+# Keep docker running easy
+exec "$@"
```

Comment:
This the only thing, that I think should also depend on environment variable. As the initialization process is long, in case when you're also processing areas, it's hard to notice the message, that everything went find and image is ready.

Maybe adding one of:
```yml
    restart: always
    restart: unless-stopped
```
In `docker-compose.yml` would solve the issue you're having, and we could remove this?

Reply:
I can't remember fully, but I'm pretty sure without that line the docker container would immediately exit after the data is finished processing. I found that pretty confusing, usually HTTP API docker container starts the service after the processing is done. also `-d` mode for `docker run` or `docker-compose up` didn't help. I'm not sure if `restart` policies would help, but could be. I can try if you want.

if your main reason is that you'd like to know when the processing is done, another alternative would be to introduce some healthcheck. that could be a curl command with some HTTP query that verifies if the container is done with data processing and started the api finally. on `docker ps` you'd see a `(starting..)` in the `STATUS` column and a `(healthy)` once the API is responsive. the API is only responsive after the area generation is done right? if you're not sure what that'd look like, you can see an example in this dockerfile over at vroom's image: https://github.com/VROOM-Project/vroom-docker/blob/10334f603215605c83ba1cf6b2c156dbbb78064c/Dockerfile#L59. that HTTP query is basically solving a tiny optimization problem every 30 seconds (default, but configurable). I'm not sure what the query with overpass would be that doesn't require knowledge about the geographic area that's loaded, but I'd hope there's some endpoint we could safely query for some sort of status. do you know a suitable endpoint?

what do you think @wiktorn ? IMO a healthcheck would be the best solution if we can find a good request.

Reply:
I guess something like that would do:
```Dockerfile
HEALTHCHECK --start-period=172800 \
    CMD curl -f http://localhost/api/status || exit 1
```

We need huge start period, so in case someone is loading planet, it will not be killed by healthcheck in the middle of the job. It would be great if we provided ability to override this value from environment

Reply:
you wouldn't exit there, it's just a check, the `--start-period` is only important if you know for certain you don't have to check the health before a certain time (here we don't really know that, but 1 mins would be fine). kartoza's postgis image shows this on `docker ps`:
```
CONTAINER ID   IMAGE                      COMMAND                  CREATED         STATUS                 PORTS                                                  NAMES
d6221f247efb   kartoza/postgis:12.1       "/bin/sh -c /docker-…"   12 months ago   Up 7 hours (healthy)   0.0.0.0:5432->5432/tcp, :::5432->5432/tcp              docker-postgis_db_1
```

so it's just adding a `healthy` in the STATUS column. if the curl is failing it'd be `starting..` instead of `healthy`, so a user would always know if the API is up and running or still processing. so I think we don't need to override the value necessarily, a failing curl every 30s is not so dramatic IMO

Reply:
but great that it has a status endpoint! makes it a lot easier

Reply:
I can add it tmrw if you agree, just ending my day:)